src/client/Microsoft.Identity.Client/OAuth2/Throttling/UiRequiredProvider.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal.Requests;
+using Microsoft.Identity.Client.PlatformsCommon.Interfaces;
+
+namespace Microsoft.Identity.Client.OAuth2.Throttling
+{
+    internal class UiRequiredProvider : IThrottlingProvider
+    {
+        /// <summary>
+        /// Default number of seconds that application returns the cached response, in case of UI required requests.
+        /// </summary>
+        internal static readonly TimeSpan s_uiRequiredExpiration = TimeSpan.FromSeconds(120);
+
+        /// <summary>
+        /// Exposed only for testing purposes
+        /// </summary>
+        internal ThrottlingCache ThrottlingCache { get; }
+
+        public UiRequiredProvider()
+        {
+            ThrottlingCache = new ThrottlingCache();
+        }
+
+        public void RecordException(AuthenticationRequestParameters requestParams, IReadOnlyDictionary<string, string> bodyParams, MsalServiceException ex)
+        {
+            if (ex is MsalUiRequiredException && IsSupported(requestParams))
+            {
+                var logger = requestParams.RequestContext.Logger;
+
+                logger.Info($"[Throttling] MsalUiRequiredException encountered - " +
+                    $"throttling for {s_uiRequiredExpiration.TotalSeconds} seconds");
+
+                var thumbprint = GetRequestStrictThumbprint(bodyParams,
+                    requestParams.AuthorityInfo.CanonicalAuthority,
+                    requestParams.RequestContext.ServiceBundle.PlatformProxy.CryptographyManager);
+                var entry = new ThrottlingCacheEntry(ex, s_uiRequiredExpiration);
+                ThrottlingCache.AddAndCleanup(thumbprint, entry, logger);
+            }
+        }
```

Comment:
can you explain when we throttle based on the ui required exception and how it's different from the regular ui exception we throw after AT S?

Reply:
Sure. The idea behind client-side throttling is to prevent developers from repeating some requests that **we know** will fail. 

Imagine the following scenario: 

- you app retries `AcquireTokenSilent` a few times. If it fails, it goes to `AcquireTokenInteractive`
- the tenant admin enables MFA
- because of MFA, existing refresh tokens become unusable
- when trying the refresh_token grant, AAD errors out with "invalid_grant" "MFA was enabled, please login interactively"
- a badly written app ignores this and simply retries with the same refresh token. Over and over. These requests are destined to fail, but they still put stress on AAD. 

With client throttling, **MSAL** detects that the application is trying to make the same bad request again and stop the request before it reaches AAD. This reduces the load on AAD, without affecting customer experience (hopefully).

Reply:
Got it. Thanks. very helpful.